planner/core/find_best_task.go

```diff
@@ -846,7 +845,105 @@ func (ds *DataSource) convertToTableScan(prop *property.PhysicalProperty, candid
 	return task, nil
 }
 
-func (ts *PhysicalTableScan) addPushedDownSelection(copTask *copTask, stats *property.StatsInfo) {
+// pushDownSelAndResolveVirtualCols push some filters down to coprocessor and resolve virtual columns.
+//	1. push down filters and check if there is virtual column.
+//	2. substitute virtual columns in TableScan's Schema to corresponding physical columns.
+//	3. substitute virtual columns in TableScan's filters and push them down.
+//	4. add a Projection upon this DataSource.
+func (ds *DataSource) pushDownSelAndResolveVirtualCols(copTask *copTask, path *accessPath, stats *property.StatsInfo) (t task) {
+	// step 1
+	t = copTask
+	if copTask.indexPlan != nil {
+		ds.addPushedDownIndexScan(copTask, path, stats)
+	}
+	if copTask.tablePlan == nil { // don't need to handle virtual columns in IndexScan
+		return
+	}
+	ts := copTask.tablePlan.(*PhysicalTableScan)
+	if !ds.hasVirtualCol {
+		ds.addPushedDownTableScan(copTask, ts, stats)
+		return
+	}
+
+	// step 2
+	ds.substituteVirtualColumns(ts)
+
+	// step 3
+	for i, expr := range ts.filterCondition {
+		ts.filterCondition[i] = expression.ColumnSubstitute(expr, ds.virtualColSchema, ds.virtualColExprs)
+	}
+	cantBePushed := ts.filterCondition // all Cast cannot be pushed down now
```

Comment:
BTW, Why all Cast cannot be pushed down now?

Reply:
`Cast` has not been supported by `Coprocesser`.

Reply:
In this PR
```sql
tidb >create table t (a int,b int as (a+1), c int, index idx(b));
Query OK, 0 rows affected
Time: 0.005s
tidb >desc select b,c from t ignore index (idx) where b>1 and c>1;
+-------------------+----------+------+------------------------------------------------------------+
| id                | count    | task | operator info                                              |
+-------------------+----------+------+------------------------------------------------------------+
| Projection_8      | 1111.11  | root | cast(plus(test.t.a, 1)), test.t.c                          |
| └─Selection_6     | 1111.11  | root | gt(cast(plus(test.t.a, 1)), 1), gt(test.t.c, 1)            |
|   └─TableReader_7 | 10000.00 | root | data:TableScan_5                                           |
|     └─TableScan_5 | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+-------------------+----------+------+------------------------------------------------------------+
```

In Master
```sql
tidb >desc select b,c from t ignore index (idx) where b>1 and c>1;
+---------------------+----------+------+------------------------------------------------------------+
| id                  | count    | task | operator info                                              |
+---------------------+----------+------+------------------------------------------------------------+
| Projection_7        | 2666.67  | root | cast(plus(test.t.a, 1)), test.t.c                          |
| └─Selection_8       | 2666.67  | root | gt(cast(plus(test.t.a, 1)), 1)                             |
|   └─TableReader_11  | 3333.33  | root | data:Selection_10                                          |
|     └─Selection_10  | 3333.33  | cop  | gt(test.t.c, 1)                                            |
|       └─TableScan_9 | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+---------------------+----------+------+------------------------------------------------------------+
```

As you can see, Master will push c>1 to tikv.

Reply:
Yes, but when `c > 1` is wrapped as `cast(c > 1 as int)`, it can't be pushed down any longer.
@lonng is working on this now, so `Cast` can be pushed down soon.

Reply:
@qw4990 `CAST` won't be cherry-picked to TiDB 3.0, any problems?

Reply:
Why `c > 1` is wrapped as `cast(c > 1 as int)`?

Reply:
My mistake, `c > 1` will not be wrapped by `cast`. 😢 
You are right that some conditions can be pushed down.
And now I use `ExpressionsToPB` to differentiate between conditions that can be pushed down and these cannot.
PTAL @crazycs520